pandas/core/exchange/column.py

```diff
@@ -0,0 +1,368 @@
+    "u": DtypeKind.UINT,
+    "f": DtypeKind.FLOAT,
+    "b": DtypeKind.BOOL,
+    "U": DtypeKind.STRING,
+    "M": DtypeKind.DATETIME,
+    "m": DtypeKind.DATETIME,
+}
+
+_NULL_DESCRIPTION = {
+    DtypeKind.FLOAT: (ColumnNullType.USE_NAN, None),
+    DtypeKind.DATETIME: (ColumnNullType.USE_NAN, None),
+    DtypeKind.INT: (ColumnNullType.NON_NULLABLE, None),
+    DtypeKind.UINT: (ColumnNullType.NON_NULLABLE, None),
+    DtypeKind.BOOL: (ColumnNullType.NON_NULLABLE, None),
+    # Null values for categoricals are stored as `-1` sentinel values
+    # in the category date (e.g., `col.values.codes` is int8 np.ndarray)
+    DtypeKind.CATEGORICAL: (ColumnNullType.USE_SENTINEL, -1),
+    # follow Arrow in using 1 as valid value and 0 for missing/null value
+    DtypeKind.STRING: (ColumnNullType.USE_BYTEMASK, 0),
+}
+
+_NO_VALIDITY_BUFFER = {
+    ColumnNullType.NON_NULLABLE: "This column is non-nullable",
+    ColumnNullType.USE_NAN: "This column uses NaN as null",
+    ColumnNullType.USE_SENTINEL: "This column uses a sentinel value",
+}
+
+
+class PandasColumn(Column):
+    """
+    A column object, with only the methods and properties required by the
+    interchange protocol defined.
+    A column can contain one or more chunks. Each chunk can contain up to three
+    buffers - a data buffer, a mask buffer (depending on null representation),
+    and an offsets buffer (if variable-size binary; e.g., variable-length
+    strings).
+    Note: this Column object can only be produced by ``__dataframe__``, so
+          doesn't need its own version or ``__column__`` protocol.
+    """
+
+    def __init__(self, column: pd.Series, allow_copy: bool = True) -> None:
+        """
+        Note: doesn't deal with extension arrays yet, just assume a regular
+        Series/ndarray for now.
+        """
+        if not isinstance(column, pd.Series):
+            raise NotImplementedError(f"Columns of type {type(column)} not handled yet")
+
+        # Store the column as a private attribute
+        self._col = column
+        self._allow_copy = allow_copy
+
+    @property
+    def size(self) -> int:
+        """
+        Size of the column, in elements.
+        """
+        return self._col.size
+
+    @property
+    def offset(self) -> int:
+        """
+        Offset of first element. Always zero.
+        """
+        # TODO: chunks are implemented now, probably this should return something
+        return 0
+
+    @cached_property
+    def dtype(self):
+        dtype = self._col.dtype
+
+        if is_categorical_dtype(dtype):
+            codes = self._col.values.codes
+            (
+                _,
+                bitwidth,
+                c_arrow_dtype_f_str,
+                _,
+            ) = self._dtype_from_pandasdtype(codes.dtype)
+            return (
+                DtypeKind.CATEGORICAL,
+                bitwidth,
+                c_arrow_dtype_f_str,
+                "=",
+            )
+        elif is_string_dtype(dtype):
```

Comment:
If a dataframe's column has object dtype, is_string_dtype returns True and the flow goes into this branch. Since the spec doesn't have a requirement to support object dtype, should we raise TypeError exception when calling df.__dataframe()?

Reply:
That's a very good question, I think I'll stub it with a `NotImplementedError` for now - I think it fits better than `TypeError` as there is no error on user side, but a missing spec entry...

Reply:
I have performed a little more research and I'm no longer sure how can I properly check for the thing being `str` vs being something more complex except checking _all_ entries in a column via `isinstance()` which feels wrong... adding a `TODO` instead.

Reply:
Why don't we check `isinstance(df.dtype, object)` in df.__dataframe() and if it is True, then throw NotImplementedError?

Reply:
because strings are usually stored as objects, don't they?.. this would effectively block strings altogether.

Reply:
Correct, strings are stored as objects. There is probably a flaw of the handling of string and object dtypes in the implementation. We should be thinking up how to handle object dtype properly.

Reply:
> i would suggest just supporting string for now and raising on object

@jreback how would I detect if a column is a string? Because I've used `is_string_dtype()` but it gives `True` on non-string objects (and I haven't found a good way to detect string-only columns except checking all their items one by one which isn't very performant).

I mean the only place I can imagine I can do better is to raise when `isinstance(obj, str)` fails during the computation of the data buffer (during the export). Is it what you mean here?

Reply:
that check is not very strict 
instead use infer_dtype which will be definitive (and fast if it's fully typed) -

Reply:
Changed in d73558a1390130a3526a5d19a48295ca0997fc61 + 4ed35bf01b2fcb5a239a71491ca1d05c41e662ef (test)

Reply:
Should this check be put in df.__dataframe__() to get an error before playing around with the dataframe implementing the protocol?

Reply:
Vasily's answer:

I cannot find this comment somewhere where I can write an answer, so I'm going to type it as a general comment.

I think this check should be delayed as much as possible because it's potentially scanning all the items in the column, which is a heavy operation while a user might just be needing some small amount of information (or might be wanting to get some particular column but not this string/object one).

Reply:
But what if the user play around with a df for a long time, which has a column with object dtype, not touching df.dtype, and only after a while gets the error. I think that is a controversial question. I would like to hear other opinions on this.

Reply:
The protocol is _mostly_ for exchanging the dataframe between certain libraries, not for some user to play around with.

I'm imagining the use case like "someone wants to plot some graphs for a few columns of a dataframe backed by library X, so they request matplotlib to show a graph; matplotlib then imports the dataframe using the protocol and shows the requested columns, it doesn't care about other columns or anything else". In this case it would be harmful to the end user of the scenario to check if any column could be represented.

Reply:
cc @jreback, @jbrockmendel, @rgommers, @dchigarev